services/shell/copy_local.go

```diff
@@ -0,0 +1,443 @@
+package shell
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"go.uber.org/multierr"
+	"go.viam.com/utils"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// copy_local supports local filesystem copy operations and is agnostic to which
+// side of the RPC copying is happening from.
+
+// NewLocalFileCopyFactory returns a FileCopyFactory that is responsible for making
+// FileCopiers that copy from the local filesystem. The destination is used later on
+// in tandem with a the CopyFilesSourceType passed into MakeFileCopier.
+func NewLocalFileCopyFactory(
+	destination string,
+	preserve bool,
+	relativeToHome bool,
+) (FileCopyFactory, error) {
+	// fixup destination to something we can work with
+	destination, err := fixPeerPath(destination, true, relativeToHome)
+	if err != nil {
+		return nil, err
+	}
+	return &localFileCopyFactory{destination: destination, preserve: preserve}, nil
+}
+
+type localFileCopyFactory struct {
+	destination string
+	preserve    bool
+}
+
+// MakeFileCopier makes a new FileCopier that is ready for to copy files into the factory's
+// file destination.
+func (f *localFileCopyFactory) MakeFileCopier(ctx context.Context, sourceType CopyFilesSourceType) (FileCopier, error) {
+	finalDestination := f.destination
+	var overrideName string
+	switch sourceType {
+	case CopyFilesSourceTypeMultipleFiles:
+		// for multiple files (a b c machine:~/some/dir), ~/some/dir neds to already exist
+		// as a directory
+		dstInfo, err := os.Stat(f.destination)
+		if err != nil || dstInfo == nil || !dstInfo.IsDir() {
+			return nil, fmt.Errorf("%q does not exist or is not a directory", f.destination)
+		}
+		if err := os.MkdirAll(filepath.Dir(f.destination), 0o750); err != nil {
+			return nil, err
+		}
+	case CopyFilesSourceTypeSingleFile, CopyFilesSourceTypeSingleDirectory:
+		// for single files (a machine:~/some/dir_or_file):
+		// if destination exists and
+		// 		it is a directory, then put the source file/directory in it.
+		//		it is a file and the source is a file, overwrrite.
```

Comment:
```suggestion
		//		it is a file and the source is a file, overwrite.
```
I'm gonna stop suggesting typo-fixes after this; they're minor nits.